src/main/java/frc/robot/subsystems/vision/VisionSubsystem.java

```diff
@@ -1,175 +1,306 @@
+  public int getNumberOfAprilTags(int limelightNumber) {
+    return limelightEstimates[limelightNumber].tagCount;
   }
 
   /**
-   * Calculates the distance between the specified robot and april tag.
-   * This method should only be called once there has been a check for if
-   * the limelights can see april tags.
+   * Returns the timestamp for the vision measurement from the limelight that is being used for pose
+   * estimation.
    */
-  private double getLimelightAprilTagDistance(int aprilTagID) {
-    if (aprilTagID >= 1 && aprilTagID <= 16) {
-      double aprilTagX = VisionConstants.APRIL_TAG_POSITIONS[aprilTagID - 1][0]; // April tag id starts at 1
-      double aprilTagY = VisionConstants.APRIL_TAG_POSITIONS[aprilTagID - 1][1];
-      Pose2d pose = getPoseFromAprilTags();
-      double robotX = pose.getX();
-      double robotY = pose.getY();
-      // Uses distance formula
-      return Math.sqrt(Math.pow(aprilTagX - robotX, 2) + Math.pow(aprilTagY - robotY, 2));
-    }
+  public double getTimeStampSeconds(int limelightNumber) {
+    return limelightEstimates[limelightNumber].timestampSeconds / 1000.0;
+  }
+
+  /**
+   * Returns the latency in seconds of when the limelight that is being used for pose estimation
+   * calculated the robot's pose. It adds the pipeline latency, capture latency, and json parsing
+   * latency.
+   */
+  public double getLatencySeconds(int limelightNumber) {
+    return (limelightEstimates[limelightNumber].latency) / 1000.0;
+  }
 
+  /**
+   * Gets the average distance between the specified limelight and the April Tags it sees
+   *
+   * @param limelightNumber the number of the limelight
+   * @return the average distance between the robot and the April Tag(s) in meters
+   */
+  public double getLimelightAprilTagDistance(int limelightNumber) {
+    if (canSeeAprilTags(limelightNumber)) {
+      return limelightEstimates[limelightNumber].avgTagDist;
+    }
     // To be safe returns a big distance from the april tags if it can't see any
     return Double.MAX_VALUE;
   }
 
+  /**
+   * Sets the heading and heading rate of the robot, this is used for deciding between MegaTag 1 and
+   * 2 for pose estimation.
+   *
+   * @param headingDegrees the angle the robot is facing in degrees (0 degrees facing the red
+   *     alliance)
+   * @param headingRateDegrees the rate the robot is rotating, CCW positive
+   */
+  public void setHeadingInfo(double headingDegrees, double headingRateDegrees) {
+    this.headingDegrees = headingDegrees;
+    this.headingRateDegreesPerSecond = headingRateDegrees;
+  }
+
+  /**
+   * Gets the limelight name associated with the specified limelight number/index
+   *
+   * @param limelightNumber the limelight number
+   * @return 0 = limelight-shooter, 1 = limelight-left, 2 = limelight-right
+   */
+  public String getLimelightName(int limelightNumber) {
+    if (limelightNumber == 0) {
```

Comment:
probably better convention to use a switch statement here, and put the exception in the default

Reply:
ok